projects/bms_carrier/test/test_passive_balance.c

```diff
@@ -0,0 +1,190 @@
+// Test sequence for passive balancing module
+#include "passive_balance.h"
+
+#include "bms_events.h"
+#include "delay.h"
+#include "interrupt.h"
+#include "ltc_afe.h"
+#include "ms_test_helpers.h"
+#include "stdio.h"
+#include "string.h"
+
+#include "log.h"
+
+// Just placeholders for afe settings from test_ltc_afe.c
+#define TEST_LTC_AFE_ADC_MODE LTC_AFE_ADC_MODE_7KHZ
+#define TEST_LTC_AFE_SPI_PORT SPI_PORT_1
+#define TEST_LTC_AFE_SPI_BAUDRATE 750000
+#define TEST_LTC_AFE_SPI_MOSI \
+  { .port = GPIO_PORT_A, .pin = 7 }
+#define TEST_LTC_AFE_SPI_MISO \
+  { .port = GPIO_PORT_A, .pin = 6 }
+#define TEST_LTC_AFE_SPI_SCLK \
+  { .port = GPIO_PORT_A, .pin = 5 }
+#define TEST_LTC_AFE_SPI_CS \
+  { .port = GPIO_PORT_A, .pin = 4 }
+
+// Number of cell voltages to be directly modified during testing
+static const uint8_t NUM_TEST_VOLTAGES = 4;
+
+static LtcAfeSettings s_test_afe_settings = {
+  .num_cells = LTC_AFE_MAX_CELLS,
+  .num_devices = LTC_AFE_MAX_DEVICES,
+  .num_thermistors = LTC_AFE_MAX_THERMISTORS,
+  .ltc_events = { .trigger_cell_conv_event = BMS_AFE_EVENT_TRIGGER_CELL_CONV,
+                  .cell_conv_complete_event = BMS_AFE_EVENT_CELL_CONV_COMPLETE,
+                  .trigger_aux_conv_event = BMS_AFE_EVENT_TRIGGER_AUX_CONV,
+                  .aux_conv_complete_event = BMS_AFE_EVENT_AUX_CONV_COMPLETE,
+                  .callback_run_event = BMS_AFE_EVENT_CALLBACK_RUN,
+                  .fault_event = BMS_AFE_EVENT_FAULT },
+  .cell_result_cb = NULL,
+  .aux_result_cb = NULL,
+  .result_context = NULL,
+  // literally random values
+  .mosi = TEST_LTC_AFE_SPI_MOSI,
+  .miso = TEST_LTC_AFE_SPI_MISO,
+  .sclk = TEST_LTC_AFE_SPI_SCLK,
+  .cs = TEST_LTC_AFE_SPI_CS,
+
+  .spi_port = TEST_LTC_AFE_SPI_PORT,
+  .spi_baudrate = TEST_LTC_AFE_SPI_BAUDRATE,
+  .adc_mode = TEST_LTC_AFE_ADC_MODE,
+};
+
+static LtcAfeStorage s_test_afe_storage;
+
+static const uint16_t AFE_DEFAULT_DISCHARGE_BITSET = 0;
+
+void setup_test(void) {
+  LOG_DEBUG("Setting up test sequence\n");
+  interrupt_init();
+  soft_timer_init();
+
+  // Configure all cells for output
+  for (uint8_t i = 0; i < LTC_AFE_MAX_DEVICES; i++) {
+    s_test_afe_settings.cell_bitset[i] = 1;
+    for (uint8_t j = 1; j < LTC_AFE_MAX_CELLS_PER_DEVICE; j++) {
+      s_test_afe_settings.cell_bitset[i] <<= 1;
+      s_test_afe_settings.cell_bitset[i] |= 0x1;
+    }
+  }
```

Comment:
Might wanna change the bitset somewhat to make sure that you end up balancing the correct one. i.e.  or 0x0F0 vs. 0xF0F